pkg/sources/chunker.go

```diff
@@ -101,39 +118,40 @@ func applyOptions(opts []ConfigOption) *chunkReaderConfig {
 }
 
 func createReaderFn(config *chunkReaderConfig) ChunkReader {
-	return func(ctx context.Context, reader io.Reader) (<-chan []byte, <-chan error) {
+	return func(ctx context.Context, reader io.Reader) <-chan ChunkResult {
 		return readInChunks(ctx, reader, config)
 	}
 }
 
-func readInChunks(ctx context.Context, reader io.Reader, config *chunkReaderConfig) (<-chan []byte, <-chan error) {
+func readInChunks(ctx context.Context, reader io.Reader, config *chunkReaderConfig) <-chan ChunkResult {
 	const channelSize = 1
 	chunkReader := bufio.NewReaderSize(reader, config.chunkSize)
-	dataChan := make(chan []byte, channelSize)
-	errChan := make(chan error, channelSize)
+	chunkResultChan := make(chan ChunkResult, channelSize)
 
 	go func() {
-		defer close(dataChan)
-		defer close(errChan)
+		defer close(chunkResultChan)
 
 		for {
+			chunkRes := ChunkResult{}
 			chunkBytes := make([]byte, config.totalSize)
 			chunkBytes = chunkBytes[:config.chunkSize]
 			n, err := chunkReader.Read(chunkBytes)
 			if n > 0 {
 				peekData, _ := chunkReader.Peek(config.totalSize - n)
 				chunkBytes = append(chunkBytes[:n], peekData...)
-				dataChan <- chunkBytes
+				chunkRes.data = chunkBytes
+				chunkResultChan <- chunkRes
 			}
 
 			if err != nil {
 				if !errors.Is(err, io.EOF) {
 					ctx.Logger().Error(err, "error reading chunk")
-					errChan <- err
+					chunkRes.err = err
+					chunkResultChan <- chunkRes
 				}
 				return
 			}
```

Comment:
If both `n > 0` and `err != nil` are true, we would send two `ChunkResult` objects:
```go
ChunkResult{data}
ChunkRelust{data, err}
```

Could that cause weird behavior in the caller?

Reply:
hmmm, it could. I'll adjust to only send one object.